### ISSUE TITLE
Windows build fails when OpenSSL not installed

### ISSUE THREAD

**Issue by Justaphf** [CauldronDevelopmentLLC/cbang#167]
Been trying to get C! to build on Windows with a very minimal dev environment (Win11, VS2022 CE, Python 3.11.3, SCons 4.8.1).  I've been running into compile errors related to not having OpenSSL installed, seems like some missing preprocessor statements from the dependency check.

I'll put in a MR once I've got it actually building successfully w/o OpenSSL, but so far these are the locations where I've encountered compile issues related to this.

`76d7bec8` - `Connection.cpp` (line 78).  `HAVE_OPENSSL` preprocessor missing from newly added line.  Not sure if the correct fix is to preprocess that line, or add an empty implementation of `setFD` to `ssh.h` when preprocessed.
<img width="407" alt="image" src="https://github.com/user-attachments/assets/3437f480-b5b0-4424-85e1-7ad77068996a">

`e91d3291` - `Port.cpp` (line 94).  `HAVE_OPENSSL` preprocessor removed but `SSLException` catch left behind unprotected by the preprocessor.  Not certain how this should be correctly fixed, but I'll plan to add the `CATCH_ERROR` macro under an #else in my MR.
<img width="826" alt="image" src="https://github.com/user-attachments/assets/6d5d5217-8b7f-4f84-96b1-dafafb90fdbb">

`632ae02f` + `29d33aea` - `Websocket.cpp` (lines 38-42 & 275) `HAVE_OPENSSL` preprocessor added around `cbang/openssl/Digest.h` include. Combined with the debug logging which requires `cbang/net/Base64.h` to be included to allow `Base64` encoding for the `writeRequest`.
<img width="712" alt="image" src="https://github.com/user-attachments/assets/91aafc1b-0be7-4ece-a35d-4e347b99e4e4">

**Comment by kbernhagen** [comment]
Maybe make a building-win document from the buildbot worker readme?

**Comment by Justaphf** [comment]
After several hours of fooling around with paths and environment variables; using the pre-built OpenSSL Windows binaries recommended by the OpenSSL wiki (https://wiki.openssl.org/index.php/Binaries  -- I used the Shining Light Win64 OpenSSL v3.4.0 MSI package), I finally got C! to build on Windows with OpenSSL and subsequently `fah-client-bastet` also built.  So far I've gotten it building, but haven't actually tried testing yet (`scons test` fails with python errors on Windows).  Definitely want to write a guide for how to configure the Windows dev environment (provided someone confirms I've done it correctly/the recommended way).  Also want to see if I can get it building OpenSSL from source instead of a pre-built binary.  Though all of that is beyond the scope of this issue which was that OpenSSL is supposed to be optional but compilation fails if OpenSSL is not installed/configured correctly.  The associated MR corrects that case, though it's not useful for FAH client builds since OpenSSL is actually required there.

@kbernhagen I looked at the `buildbot` project you suggested for about 30 minutes but didn't see anything quickly relevant to setting up the Windows environment with OpenSSL working.  The `README.w32` was last updated 19 years ago.  Let me know if I was looking in the wrong place (https://github.com/CauldronDevelopmentLLC/buildbot/blob/0.7.10p2-jcoffland/README.w32)

To get it compiling with OpenSSL I needed to:
1. Install OpenSSL (64-bit) pre-built for Windows.
2. Configure environment variables (not sure all of these are needed but this is what I had when it finally started working): `OPENSSL_DIR`, `OPENSSL_HOME`, `OPENSSL_INCLUDE_DIR`, `OPENSSL_LIB_DIR`
3. Copy the MT files from `C:\Program Files\OpenSSL-Win64\lib\VC\x64\MT` to `C:\Program Files\OpenSSL-Win64\lib`.  This is likely due to the installation structure for the particular pre-build OpenSSL.
4. Either set additional environment variables or create an `env.bat` file similar to the instructions from (https://github.com/CauldronDevelopmentLLC/CAMotics/blob/master/doc/Windows_Build.md) (TARGET_ARCH=x86_64, CBANG_HOME=git checkout folder for cbang -- needed for the fah-client-bastet, not necessary for cbang itself).

**Comment by kbernhagen** [comment]
I was referring to a private repo `fah-build`, asking Joseph.
The custom buildbot repo you looked at is obsolete.

The latest buildbot was used with from-json-config scripts similar to what is in camotics. But the Win instructions are very different.

I believe I built Strawberry perl to build OpenSSL.
I think OpenSSL 3.0 is currently used. Don't know if 3.4 is ok with fah.

**Comment by kbernhagen** [comment]
I believe openssl is statically linked in the client on macOS and Windows.

**Comment by Justaphf** [comment]
I was using the MT libraries for static linking but was missing the `OPENSSL_STATIC=1` environment variable.  Not sure if that affected anything with the build but it still built without issues. I also got `fah-client-bastet` packaging to work on Windows after installing NSIS, however I haven't actually attempted to install/run it as my dev machine is currently involved in folding month for another week so don't want to mess with that until after that ends.

Thanks for the quick feedback.  I'll look into things more tomorrow.

**Comment by kbernhagen** [comment]
There is probably supposed to be `mostly_static=1` added to the scons build commands.

**Comment by Justaphf** [comment]
Thanks for the help.  I think I've figured out minimal environment variables to set such that the release build compiles on my machine. However, the tests don't work when C! is compiled in debug mode.  Seems like the debug flag passed at the command line isn't propagated to the tests when they are built so I get linker errors trying to link against the debug libraries.  I can get the tests to build in debug mode if I scons directly in the tests folder, but then if I try to run the tests from there they can't find some needed files. If I build from below the tests location and then try to run from the top level again it of course goes back to trying to rebuild the tests as release and fails linking.

I just tested in Linux and seems like it has the same issue.  If I build debug and then try to test I get linker failures (regardless if I set the `debug=1` on the command line for tests).  If I build release and run tests under release everything passes in Linux.

Running the tests in release mode on Windows does have other issues I'm still trying to overcome.  I suspect there is something missing or incorrectly configured with my Python install.  Starting with the `ioStreamTests` I get failures related to stdout being blank when something was expected to be returned.  Compression tests pass then starting with the `ipRangeTests` the rest of the tests fail up until the UTF8 test crashes with
```UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 4: character maps to <undefined>```

For reference here is my script that is working up to the test failures mentioned.  Still using the pre-built 64-bit OpenSSL 3.4.0 so not sure if that's related to any of my issues.
```
call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64

SET TARGET_ARCH=amd64
SET "OPENSSL_INCLUDE=C:\Program Files\OpenSSL-Win64\include"
:: For debug use MTd, for release use MT
::SET "OPENSSL_LIBPATH=C:\Program Files\OpenSSL-Win64\lib\VC\x64\MTd"
SET "OPENSSL_LIBPATH=C:\Program Files\OpenSSL-Win64\lib\VC\x64\MT"
SET CBANG_HOME=C:\Development\cbang

cd C:\Development
scons -C cbang --clean
scons -C cbang mostly_static=1
scons -C cbang mostly_static=1 test
```

These are the packages I have installed in Python:
```
Package         Version
--------------- -------
confluent-kafka 2.1.1
pip             24.3.1
pywin32         306
SCons           4.8.1
setuptools      65.5.0
six             1.16.0
vboxapi         1.0
vboxapi         1.0
vboxapi         1.0
wheel           0.40.0
```
Also note I tested against `master` rather than the MR I added so test failures shouldn't be related to that.
